### PR TITLE
Feature/2016 05 01 andreas evers jwt for state

### PR DESCRIPTION
**Pull request by andreasevers**

### CODE REVIEW

File: _posts/2016-05-01-Using-JWT-Tokens-for-State-Transfer.md [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+comments: true
+---
+
+At one of our clients, we have been using JWT tokens quite extensibly.
+We even use it to persist state on the client.
+
+## Why persisting state on the client?
+
+When buiding microservices, we need to build so-called "cloud native" applications.
+One of the key tenents of cloud native application design is keeping your services stateless.
```
Comment (tomverelst) [inline _posts/2016-05-01-Using-JWT-Tokens-for-State-Transfer.md]: key tenets

File: _posts/2016-05-01-Using-JWT-Tokens-for-State-Transfer.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+The benefit of having stateless applications is foremost the ability to respond to events by adding or removing instances without needing to significantly reconfigure or change the application.
+More stateless services can easily be added when load suddenly increases, or if an existing stateless service fails, it can simply be replaced with another.
+Hence, resilience and agility, are easier to achieve with stateless services.
+
+Keeping your services stateless means we need to persist our state somewhere else.
+Since we are tranferring state in the REST architectural style, we can use the client to retain our state.
```
Comment (bart-blommaerts) [inline _posts/2016-05-01-Using-JWT-Tokens-for-State-Transfer.md]: in **a** REST architectural style?

File: _posts/2016-05-01-Using-JWT-Tokens-for-State-Transfer.md [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+For scaling purposes this is a great solution, as the client will only ever have to store its own state, and the server will be relieved of the state of all its clients.
+
+At our client we have chosen to use JWT tokens for this state transfer to the client.
+While JWT is primarily intended for authentication and authorization purposes, the specification allows us to add any data we'd like to verify later on.
+
+## Looking good
+
+Imagine the following scenario:
+
+A list of products is fetched from the products microservice.
```
Comment (bart-blommaerts) [inline _posts/2016-05-01-Using-JWT-Tokens-for-State-Transfer.md]: +A list of products is fetched from the "products microservice".

File: _posts/2016-05-01-Using-JWT-Tokens-for-State-Transfer.md [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+The user isn't allowed to view all products, so only those products the user has access to are returned.
+When the user wants to order a product, he sends an order request to the order microservice with the id of the product he wants to order.
```
Comment (bart-blommaerts) [inline _posts/2016-05-01-Using-JWT-Tokens-for-State-Transfer.md]: he sends an order request to the "orders microservice"

File: _posts/2016-05-01-Using-JWT-Tokens-for-State-Transfer.md [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+At that moment the orders microservice needs to know whether or not the user is allowed to order access this product, let alone order it.
```
Comment (bart-blommaerts) [inline _posts/2016-05-01-Using-JWT-Tokens-for-State-Transfer.md]: +At that moment the "orders microservice"

And so on for all "identifying" applications.

File: _posts/2016-05-01-Using-JWT-Tokens-for-State-Transfer.md [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+Since the rights to access and order are the same, we'd like to reuse the information returned from the first call to the products microservice.
+This flow is illustrated below.
+
+![JWT state transfer]({{ '/img/JWT/jwt-for-state-tranfer.png' | prepend: site.baseurl }})
+
+We could call the products microservice from the orders microservice and count on caching, but that would still be an extra network hop and the cache could potentially be invalidated by the time the user orders the product.
+Using the JWT approach, state is given to the client (the list of product ids the user is allowed to access), and being passed to the server again the moment an order is placed.
+The signature of the JWT token guarantees us that the state is not tampered with while residing on the client.
```
Comment (bart-blommaerts) [inline _posts/2016-05-01-Using-JWT-Tokens-for-State-Transfer.md]: The signature of the JWT token guarantees us that the state is not tampered with, while residing on the client

File: _posts/2016-05-01-Using-JWT-Tokens-for-State-Transfer.md [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+Yet by transferring the JWT token from one microservice over the client to the other microservice, we are creating a hidden dependency.
+It's also hard to have versioning on headers unless we put the version inside the name of the header.
+
+### Scaling
+
+Adding versions to the headernames, documenting which microservices expect which versions of JWT tokens of other microservices, and making sure we implement the tolerant-reader principle when reading the tokens might be a step in the right direction to avoid mass hysteria when tokens have to be adjusted.
+But what is simply impossible to get around is the size restriction of headers in HTTP requests and responses.
+The HTTP specification doesn't put any restriction on header size (singular or combined).
+But webservers, reverse proxies, CDNs and other network components do.
+Why they do this is not entirely clear as the spec allows any size, but the fact of the matter is that the restrictions exist.
```
Comment (bart-blommaerts) [inline _posts/2016-05-01-Using-JWT-Tokens-for-State-Transfer.md]: that these restrictions exist.

File: _posts/2016-05-01-Using-JWT-Tokens-for-State-Transfer.md [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+The second alternative is to persist the state in a key value datastore on the server.
+We could asynchronously fetch products data and store it inside a datastore owned by the orders microservice.
+This could get stale, but so could a cache on the products microservice.
+This approach seems most common in the industry and could be well be the most preferrable.
+
+And when all else fails, we can still simply make a call from the orders microservice to the products microservice and count on caching.
+
+## Conclusion
+
+Using JWT tokens as a means to transfer state to and from microservices via the client could seem like a good idea, but in the end turned out to be quite an anti-pattern.
```
Comment (bart-blommaerts) [inline _posts/2016-05-01-Using-JWT-Tokens-for-State-Transfer.md]: Using JWT tokens as a means to transfer state to and from microservices via the client seemed like a good idea

**Comment by bart-blommaerts** [comment]
Interessante post. Misschien een follow-up met de "shared" data oplossing?

**Pull request merged**